File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/resolver/RemoteRecordsResolver.java
package ru.citeck.ecos.records3.record.resolver;

import ecos.com.fasterxml.jackson210.databind.node.ObjectNode;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.commons.json.JsonMapper;
import ru.citeck.ecos.commons.utils.StringUtils;
import ru.citeck.ecos.records2.RecordMeta;
import ru.citeck.ecos.records2.request.delete.RecordsDelResult;
import ru.citeck.ecos.records2.request.mutation.RecordsMutResult;
import ru.citeck.ecos.records2.request.query.typed.RecordsMetaQueryResult;
import ru.citeck.ecos.records2.request.rest.DeletionBody;
import ru.citeck.ecos.records2.request.rest.MutationBody;
import ru.citeck.ecos.records2.rest.RemoteRecordsRestApi;
import ru.citeck.ecos.records3.record.op.atts.dto.RecordAtts;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.RecordsServiceFactory;
import ru.citeck.ecos.records3.record.op.delete.dto.DelStatus;
import ru.citeck.ecos.records3.record.request.RequestContext;
import ru.citeck.ecos.records3.record.request.RequestCtxData;
import ru.citeck.ecos.records3.record.request.msg.MsgLevel;
import ru.citeck.ecos.records3.rest.v1.RequestBody;
import ru.citeck.ecos.records3.rest.v1.delete.DeleteBody;
import ru.citeck.ecos.records3.rest.v1.delete.DeleteResp;
import ru.citeck.ecos.records3.rest.v1.mutate.MutateBody;
import ru.citeck.ecos.records3.record.op.query.dto.RecordsQuery;
import ru.citeck.ecos.records3.record.op.query.dto.RecsQueryRes;
import ru.citeck.ecos.records3.rest.v1.mutate.MutateResp;
import ru.citeck.ecos.records3.rest.v1.query.QueryBody;
import ru.citeck.ecos.records3.rest.v1.query.QueryResp;
import ru.citeck.ecos.records3.record.dao.RecordsDaoInfo;
import ru.citeck.ecos.records2.utils.RecordsUtils;
import ru.citeck.ecos.records2.utils.ValWithIdx;
import ru.citeck.ecos.records3.utils.V1ConvUtils;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Slf4j
public class RemoteRecordsResolver {

    public static final String BASE_URL = "/api/records/";
    public static final String QUERY_URL = BASE_URL + "query";
    public static final String MUTATE_URL = BASE_URL + "mutate";
    public static final String DELETE_URL = BASE_URL + "delete";

    private final RemoteRecordsRestApi restApi;
    private String defaultAppName = "";

    private final Map<String, String> sourceIdMapping = new HashMap<>();

    private final JsonMapper mapper = Json.getMapper();

    public RemoteRecordsResolver(RecordsServiceFactory factory, RemoteRecordsRestApi restApi) {
        this.restApi = restApi;

        Map<String, String> sourceIdMapping = factory.getProperties().getSourceIdMapping();
        if (sourceIdMapping != null) {
            this.sourceIdMapping.putAll(sourceIdMapping);
        }
    }

    public RecsQueryRes<RecordAtts> query(@NotNull RecordsQuery query,
                                          @NotNull Map<String, ?> attributes,
                                          boolean rawAtts) {

        RequestContext context = RequestContext.getCurrentNotNull();

        String sourceId = query.getSourceId();
        if (sourceId.indexOf('/') == -1) {
            sourceId = defaultAppName + "/" + sourceId;
        }

        sourceId = sourceIdMapping.getOrDefault(sourceId, sourceId);

        String appName;

        RecordsQuery appQuery = new RecordsQuery(query);

        int appDelimIdx = sourceId.indexOf("/");
        appName = sourceId.substring(0, appDelimIdx);
        appQuery.setSourceId(sourceId.substring(appDelimIdx + 1));

        QueryBody queryBody = new QueryBody();

        queryBody.setQuery(appQuery);
        queryBody.setAttributes(attributes);
        queryBody.setRawAtts(rawAtts);
        setContextProps(queryBody, context);

        QueryResp queryResp = execQuery(appName, queryBody, context);
        RecsQueryRes<RecordAtts> result = new RecsQueryRes<>();
        result.setRecords(queryResp.getRecords());
        result.setTotalCount(queryResp.getTotalCount());
        result.setHasMore(queryResp.isHasMore());

        return RecordsUtils.attsWithDefaultApp(result, appName);
    }

    @NotNull
    private QueryResp execQuery(String appName, QueryBody queryBody, RequestContext context) {

        ru.citeck.ecos.records2.request.rest.QueryBody v0Body = toV0QueryBody(queryBody, context);

        ObjectNode appResultObj = postRecords(appName, QUERY_URL, v0Body);
        QueryResp result = toQueryAttsRes(appResultObj, context);
        if (result == null) {
            result = new QueryResp();
        } else {
            context.addAllMsgs(result.getMessages());
        }

        return result;
    }

    @Nullable
    private QueryResp toQueryAttsRes(ObjectNode body, RequestContext context) {

        if (body == null || body.isEmpty()) {
            return null;
        }

        if (body.path("version").asInt(0) == 1) {
            return mapper.convert(body, QueryResp.class);
        }

        RecordsMetaQueryResult v0Result = mapper.convert(body, RecordsMetaQueryResult.class);
        if (v0Result == null) {
            return null;
        }

        V1ConvUtils.addErrorMessages(v0Result.getErrors(), context);
        V1ConvUtils.addDebugMessage(v0Result, context);

        QueryResp resp = new QueryResp();
        resp.setRecords(v0Result.getRecords()
            .stream()
            .map(RecordAtts::new)
            .collect(Collectors.toList()));
        resp.setHasMore(v0Result.getHasMore());
        resp.setTotalCount(v0Result.getTotalCount());

        return resp;
    }

    private ru.citeck.ecos.records2.request.rest.QueryBody toV0QueryBody(QueryBody body, RequestContext context) {

        ru.citeck.ecos.records2.request.rest.QueryBody v0Body = new ru.citeck.ecos.records2.request.rest.QueryBody();
        v0Body.setAttributes(body.getAttributes());
        v0Body.setRecords(body.getRecords());
        v0Body.setV1Body(body);

        if (body.getQuery() != null) {
            v0Body.setQuery(V1ConvUtils.recsQueryV1ToV0(body.getQuery(), context));
        }

        return v0Body;
    }

    public List<RecordAtts> getAtts(@NotNull List<RecordRef> records,
                                    @NotNull Map<String, ?> attributes,
                                    boolean rawAtts) {

        RequestContext context = RequestContext.getCurrentNotNull();

        List<ValWithIdx<RecordAtts>> result = new ArrayList<>();

        Map<String, List<ValWithIdx<RecordRef>>> refsByApp = RecordsUtils.groupByApp(records);
        refsByApp.forEach((app, refs) -> {

            if (StringUtils.isBlank(app)) {
                app = defaultAppName;
            }
            QueryBody queryBody = new QueryBody();
            queryBody.setRecords(refs.stream().map(ValWithIdx::getValue).collect(Collectors.toList()));
            queryBody.setAttributes(attributes);
            queryBody.setRawAtts(rawAtts);

            QueryResp queryResp = execQuery(app, queryBody, context);

            if (queryResp.getRecords() == null || queryResp.getRecords().size() != refs.size()) {
                log.error("Incorrect response: " + queryBody + "\n query: " + queryBody);
                for (ValWithIdx<RecordRef> ref : refs) {
                    result.add(new ValWithIdx<>(new RecordAtts(ref.getValue()), ref.getIdx()));
                }
            } else {
                List<RecordAtts> recsAtts = queryResp.getRecords();
                for (int i = 0; i < refs.size(); i++) {
                    ValWithIdx<RecordRef> ref = refs.get(i);
                    RecordAtts atts = recsAtts.get(i);
                    result.add(new ValWithIdx<>(new RecordAtts(atts, ref.getValue()), ref.getIdx()));
                }
            }
        });

        result.sort(Comparator.comparingInt(ValWithIdx::getIdx));
        return result.stream().map(ValWithIdx::getValue).collect(Collectors.toList());
    }

    public List<RecordRef> mutate(@NotNull List<RecordAtts> records) {

        RequestContext context = RequestContext.getCurrentNotNull();
        List<ValWithIdx<RecordRef>> result = new ArrayList<>();

        Map<String, List<ValWithIdx<RecordAtts>>> attsByApp = RecordsUtils.groupAttsByApp(records);

        attsByApp.forEach((app, atts) -> {

            if (StringUtils.isBlank(app)) {
                app = defaultAppName;
            }

            MutateBody mutateBody = new MutateBody();
            mutateBody.setRecords(atts.stream()
                .map(ValWithIdx::getValue)
                .collect(Collectors.toList()));

            setContextProps(mutateBody, context);

            MutationBody v0Body = new MutationBody();
            if (context.isMsgEnabled(MsgLevel.DEBUG)) {
                v0Body.setDebug(true);
            }
            v0Body.setRecords(mapper.convert(mutateBody.getRecords(), mapper.getListType(RecordMeta.class)));
            v0Body.setV1Body(mutateBody);

            ObjectNode mutRespObj = postRecords(app, MUTATE_URL, v0Body);
            MutateResp mutateResp = toMutateResp(mutRespObj, context);

            if (mutateResp != null && mutateResp.getMessages() != null) {
                mutateResp.getMessages().forEach(context::addMsg);
            }

            if (mutateResp == null
                    || mutateResp.getRecords() == null
                    || mutateResp.getRecords().size() != atts.size()) {

                context.addMsg(MsgLevel.ERROR, () -> "Incorrect response: " + mutateResp + "\n query: " + mutateBody);
                for (ValWithIdx<RecordAtts> att : atts) {
                    result.add(new ValWithIdx<>(att.getValue().getId(), att.getIdx()));
                }
            } else {
                List<RecordAtts> recsAtts = mutateResp.getRecords();
                for (int i = 0; i < atts.size(); i++) {
                    ValWithIdx<RecordAtts> refAtts = atts.get(i);
                    RecordAtts respAtts = recsAtts.get(i);
                    result.add(new ValWithIdx<>(respAtts.getId(), refAtts.getIdx()));
                }
            }
        });

        result.sort(Comparator.comparingInt(ValWithIdx::getIdx));
        return result.stream().map(ValWithIdx::getValue).collect(Collectors.toList());
    }

    @Nullable
    private MutateResp toMutateResp(ObjectNode body, RequestContext context) {

        if (body == null || body.isEmpty()) {
            return null;
        }

        if (body.path("version").asInt(0) == 1) {
            return mapper.convert(body, MutateResp.class);
        }

        RecordsMutResult v0Result = mapper.convert(body, RecordsMutResult.class);
        if (v0Result == null) {
            return null;
        }

        V1ConvUtils.addErrorMessages(v0Result.getErrors(), context);
        V1ConvUtils.addDebugMessage(v0Result, context);

        MutateResp resp = new MutateResp();
        resp.setRecords(v0Result.getRecords()
            .stream()
            .map(RecordAtts::new)
            .collect(Collectors.toList()));

        return resp;
    }

    private RecordAtts addAppName(RecordAtts meta, String app) {
        return new RecordAtts(meta, r -> r.addAppName(app));
    }

    private RecordAtts removeAppName(RecordAtts meta) {
        return new RecordAtts(meta, RecordRef::removeAppName);
    }

    @NotNull
    public List<DelStatus> delete(@NotNull List<RecordRef> records) {

        RequestContext context = RequestContext.getCurrentNotNull();
        List<ValWithIdx<DelStatus>> result = new ArrayList<>();

        Map<String, List<ValWithIdx<RecordRef>>> attsByApp = RecordsUtils.groupByApp(records);
        attsByApp.forEach((app, refs) -> {

            if (StringUtils.isBlank(app)) {
                app = defaultAppName;
            }

            DeleteBody deleteBody = new DeleteBody();
            deleteBody.setRecords(refs.stream()
                .map(ValWithIdx::getValue)
                .collect(Collectors.toList()));
            setContextProps(deleteBody, context);

            DeletionBody v0Body = new DeletionBody();
            v0Body.setRecords(deleteBody.getRecords());

            if (context.isMsgEnabled(MsgLevel.DEBUG)) {
                v0Body.setDebug(true);
            }
            v0Body.setV1Body(deleteBody);

            ObjectNode delRespObj = postRecords(app, DELETE_URL, v0Body);
            DeleteResp resp = toDeleteResp(delRespObj, context);

            if (resp != null && resp.getMessages() != null) {
                resp.getMessages().forEach(context::addMsg);
            }

            List<DelStatus> statues = toDelStatuses(refs.size(), resp, context);

            for (int i = 0; i < refs.size(); i++) {
                ValWithIdx<RecordRef> refAtts = refs.get(i);
                DelStatus status = statues.get(i);
                result.add(new ValWithIdx<>(status, refAtts.getIdx()));
            }
        });

        result.sort(Comparator.comparingInt(ValWithIdx::getIdx));
        return result.stream().map(ValWithIdx::getValue).collect(Collectors.toList());
    }

    @Nullable
    private DeleteResp toDeleteResp(ObjectNode data, RequestContext context) {

        if (data == null || data.size() == 0) {
            return new DeleteResp();
        }

        if (data.path("version").asInt(0) == 1) {
            return mapper.convert(data, DeleteResp.class);
        }

        RecordsDelResult v0Resp = mapper.convert(data, RecordsDelResult.class);
        if (v0Resp == null) {
            return new DeleteResp();
        }
        List<RecordMeta> records = v0Resp.getRecords();

        DeleteResp resp = new DeleteResp();
        resp.setStatuses(records.stream()
            .map(r -> DelStatus.OK)
            .collect(Collectors.toList()));

        V1ConvUtils.addErrorMessages(v0Resp.getErrors(), context);
        V1ConvUtils.addDebugMessage(v0Resp, context);

        return resp;
    }

    private List<DelStatus> toDelStatuses(int expectedSize, DeleteResp resp, RequestContext context) {

        if (resp == null) {
            return getDelStatuses(expectedSize, DelStatus.ERROR);
        }

        if (resp.getStatuses() != null && resp.getStatuses().size() == expectedSize) {
            return resp.getStatuses();
        }

        context.addMsg(MsgLevel.ERROR, () ->
            "Result statues doesn't match request. "
                + "Expected size: " + expectedSize
                + ". Actual response: " + resp );

        return getDelStatuses(expectedSize, DelStatus.ERROR);
    }

    private List<DelStatus> getDelStatuses(int size, DelStatus status) {
        return Stream.generate(() -> status)
            .limit(size)
            .collect(Collectors.toList());
    }

    private void setContextProps(RequestBody body, RequestContext ctx) {
        RequestCtxData<?> ctxData = ctx.getCtxData();
        body.setMsgLevel(ctxData.getMsgLevel());
        body.setRequestId(ctxData.getRequestId());
        body.setRequestTrace(ctxData.getRequestTrace());
    }

    @Nullable
    private ObjectNode postRecords(String appName, String url, Object body) {
        String appUrl = "/" + appName + url;
        return restApi.jsonPost(appUrl, body, ObjectNode.class);
    }

    @Nullable
    public RecordsDaoInfo getSourceInfo(@NotNull String sourceId) {
        //todo
        return null;
    }

    @NotNull
    public List<RecordsDaoInfo> getSourceInfo() {
        //todo
        return Collections.emptyList();
    }

    public void setDefaultAppName(String defaultAppName) {
        this.defaultAppName = defaultAppName;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/resolver/RemoteRecordsResolver.kt
package ru.citeck.ecos.records3.record.resolver

import mu.KotlinLogging
import ru.citeck.ecos.commons.json.Json
import ru.citeck.ecos.commons.utils.StringUtils
import ru.citeck.ecos.records2.RecordRef
import ru.citeck.ecos.records2.exception.RecordsException
import ru.citeck.ecos.records2.exception.RemoteRecordsException
import ru.citeck.ecos.records2.request.error.RecordsError
import ru.citeck.ecos.records2.utils.RecordsUtils
import ru.citeck.ecos.records2.utils.ValWithIdx
import ru.citeck.ecos.records3.RecordsService
import ru.citeck.ecos.records3.RecordsServiceFactory
import ru.citeck.ecos.records3.cache.Cache
import ru.citeck.ecos.records3.cache.CacheConfig
import ru.citeck.ecos.records3.record.atts.dto.RecordAtts
import ru.citeck.ecos.records3.record.atts.schema.annotation.AttName
import ru.citeck.ecos.records3.record.dao.delete.DelStatus
import ru.citeck.ecos.records3.record.dao.impl.source.RecordsSourceMeta
import ru.citeck.ecos.records3.record.dao.query.dto.query.RecordsQuery
import ru.citeck.ecos.records3.record.dao.query.dto.res.RecsQueryRes
import ru.citeck.ecos.records3.record.request.RequestContext
import ru.citeck.ecos.records3.record.request.msg.MsgLevel
import ru.citeck.ecos.records3.record.request.msg.ReqMsg
import ru.citeck.ecos.records3.rest.v1.RequestBody
import ru.citeck.ecos.records3.rest.v1.RequestResp
import ru.citeck.ecos.records3.rest.v1.delete.DeleteBody
import ru.citeck.ecos.records3.rest.v1.delete.DeleteResp
import ru.citeck.ecos.records3.rest.v1.mutate.MutateBody
import ru.citeck.ecos.records3.rest.v1.mutate.MutateResp
import ru.citeck.ecos.records3.rest.v1.query.QueryBody
import ru.citeck.ecos.records3.rest.v1.query.QueryResp
import ru.citeck.ecos.records3.rest.v1.txn.TxnBody
import ru.citeck.ecos.records3.rest.v1.txn.TxnResp
import ru.citeck.ecos.records3.rest.v2.query.QueryBodyV2
import ru.citeck.ecos.records3.security.HasSensitiveData
import ru.citeck.ecos.webapp.api.web.EcosWebClient
import java.util.concurrent.TimeUnit
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.reflect.KClass

class RemoteRecordsResolver(
    val services: RecordsServiceFactory
) {

    companion object {
        val log = KotlinLogging.logger {}

        const val BASE_PATH: String = "/records/"
        const val QUERY_PATH: String = BASE_PATH + "query"
        const val MUTATE_PATH: String = BASE_PATH + "mutate"
        const val DELETE_PATH: String = BASE_PATH + "delete"
        const val TXN_PATH: String = BASE_PATH + "txn"
    }

    private var defaultAppName: String = services.properties.defaultApp
    private val sourceIdMapping = services.properties.sourceIdMapping
    private lateinit var recordsService: RecordsService
    private val txnActionManager = services.txnActionManager

    private val sourceIdMeta: Cache<String, RecSrcMeta>
    private val webClient: EcosWebClient = services.getEcosWebAppContext()?.getWebClient()
        ?: error("EcosWebAppContext or WebClient is null")

    init {
        sourceIdMeta = services.cacheManager.create(
            CacheConfig(
                key = "remote-source-id-meta",
                expireAfterWrite = TimeUnit.MINUTES.toMillis(1),
                maxItems = 200
            ),
            RecSrcMeta(false)
        ) { k -> evalSourceIdMeta(k) }
    }

    fun query(
        query: RecordsQuery,
        attributes: Map<String, *>,
        rawAtts: Boolean
    ): RecsQueryRes<RecordAtts> {

        val context: RequestContext = RequestContext.getCurrentNotNull()
        var sourceId = query.sourceId

        if (sourceId.indexOf('/') == -1) {
            sourceId = "$defaultAppName/$sourceId"
        }
        sourceId = sourceIdMapping.getOrDefault(sourceId, sourceId)
        val appName: String

        val appDelimIdx = sourceId.indexOf("/")
        appName = sourceId.substring(0, appDelimIdx)

        val appQuery = query.copy()
            .withSourceId(sourceId.substring(appDelimIdx + 1))
            .build()

        val queryBody = if (webClient.getApiVersion(appName, QUERY_PATH) >= 2) {
            QueryBodyV2()
        } else {
            QueryBody()
        }
        queryBody.setQuery(appQuery)
        queryBody.setAttributes(attributes)
        queryBody.rawAtts = rawAtts
        setContextProps(queryBody, context)

        val queryResp = exchangeRemoteRequest(appName, QUERY_PATH, queryBody, QueryResp::class, context)

        val result = RecsQueryRes<RecordAtts>()

        result.setRecords(queryResp.records)
        result.setTotalCount(queryResp.totalCount)
        result.setHasMore(queryResp.hasMore)

        return RecordsUtils.attsWithDefaultApp(result, appName)
    }

    fun getAtts(records: List<RecordRef>, attributes: Map<String, *>, rawAtts: Boolean): List<RecordAtts> {

        val context: RequestContext = RequestContext.getCurrentNotNull()
        val result = ArrayList<ValWithIdx<RecordAtts>>()
        val refsByApp = RecordsUtils.groupByApp(records)

        refsByApp.forEach { (appArg, refs) ->

            val app = if (StringUtils.isBlank(appArg)) {
                defaultAppName
            } else {
                appArg
            }

            val recsAtts: List<RecordAtts> = getAttsForApp(
                app,
                refs.map { it.value.removeAppName() },
                attributes,
                rawAtts,
                context
            )
            for (i in refs.indices) {
                val ref: ValWithIdx<RecordRef> = refs[i]
                val atts: RecordAtts = recsAtts[i]
                result.add(ValWithIdx(RecordAtts(atts, ref.value), ref.idx))
            }
        }
        result.sortBy { it.idx }
        return result.map { it.value }
    }

    private fun getAttsForApp(
        appName: String,
        records: List<RecordRef>,
        attributes: Map<String, *>,
        rawAtts: Boolean,
        context: RequestContext
    ): List<RecordAtts> {

        val queryBody = if (webClient.getApiVersion(appName, QUERY_PATH) >= 2) {
            QueryBodyV2()
        } else {
            QueryBody()
        }
        queryBody.setRecords(records)
        queryBody.setAttributes(attributes)
        queryBody.rawAtts = rawAtts
        setContextProps(queryBody, context)

        val queryResp = exchangeRemoteRequest(appName, QUERY_PATH, queryBody, QueryResp::class, context)

        if (queryResp.records.size != records.size) {
            throw RecordsException(
                "Incorrect " +
                    "response: ${formatObjForLog(queryResp)} " +
                    "query: ${formatObjForLog(queryBody)}"
            )
        }

        return queryResp.records
    }

    // todo: records grouping by appName performed on RecordsService level and this method should be refactored
    fun mutate(records: List<RecordAtts>, attsToLoad: Map<String, *>, rawAtts: Boolean): List<RecordAtts> {

        val context: RequestContext = RequestContext.getCurrentNotNull()
        val result: MutableList<ValWithIdx<RecordAtts>> = ArrayList()
        var attsByApp: Map<String, List<ValWithIdx<RecordAtts>>> = RecordsUtils.groupAttsByApp(records)

        if (defaultAppName.isNotEmpty() && attsByApp.containsKey(defaultAppName) && attsByApp.containsKey("")) {
            val newDefaultAppRecs = ArrayList(attsByApp[defaultAppName] ?: emptyList())
            newDefaultAppRecs.addAll(attsByApp[""] ?: emptyList())
            newDefaultAppRecs.sortBy { it.idx }
            val newAttsByApp = HashMap(attsByApp)
            newAttsByApp[defaultAppName] = newDefaultAppRecs
            newAttsByApp.remove("")
            attsByApp = newAttsByApp
        }

        attsByApp.forEach { (appArg, atts) ->

            val appName = if (StringUtils.isBlank(appArg)) {
                defaultAppName
            } else {
                appArg
            }

            val mutateBody = MutateBody()
            mutateBody.setRecords(atts.map { it.value.withoutAppName() })
            mutateBody.setAttributes(attsToLoad)
            mutateBody.rawAtts = rawAtts

            setContextProps(mutateBody, context)

            val mutateResp: MutateResp = exchangeRemoteRequest(
                appName,
                MUTATE_PATH,
                mutateBody,
                MutateResp::class,
                context
            )
            if (mutateResp.records.size != atts.size) {
                throw RecordsException(
                    "Incorrect " +
                        "response: ${formatObjForLog(mutateResp)} " +
                        "query: ${formatObjForLog(mutateBody)}"
                )
            } else {
                val recsAtts = mutateResp.records
                for (i in atts.indices) {
                    val refAtts = atts[i]
                    val newAtts = recsAtts[i].withDefaultAppName(appName)
                    result.add(ValWithIdx(newAtts, refAtts.idx))
                }
            }
        }
        result.sortBy { it.idx }
        return result.map { it.value }
    }

    fun delete(records: List<RecordRef>): List<DelStatus> {

        val context: RequestContext = RequestContext.getCurrentNotNull()
        val result: MutableList<ValWithIdx<DelStatus>> = ArrayList()
        val attsByApp: Map<String, MutableList<ValWithIdx<RecordRef>>> = RecordsUtils.groupByApp(records)

        attsByApp.forEach { (appArg, refs) ->

            val app = if (StringUtils.isBlank(appArg)) {
                defaultAppName
            } else {
                appArg
            }

            val deleteBody = DeleteBody()
            deleteBody.setRecords(refs.map { it.value.removeAppName() })
            setContextProps(deleteBody, context)

            val resp: DeleteResp = exchangeRemoteRequest(app, DELETE_PATH, deleteBody, DeleteResp::class, context)

            val statuses = resp.statuses
            if (statuses.size != deleteBody.records.size) {
                throw RecordsException(
                    "Result statues doesn't match request. " +
                        "Expected size: " + deleteBody.records.size +
                        ". Actual response: " + formatObjForLog(resp)
                )
            }
            for (i in refs.indices) {
                val refAtts: ValWithIdx<RecordRef> = refs[i]
                result.add(ValWithIdx(statuses[i], refAtts.idx))
            }
        }
        result.sortBy { it.idx }
        return result.map { it.value }
    }

    fun commit(recordRefs: List<RecordRef>) {
        completeTransaction(recordRefs, TxnBody.TxnAction.COMMIT)
    }

    fun rollback(recordRefs: List<RecordRef>) {
        completeTransaction(recordRefs, TxnBody.TxnAction.ROLLBACK)
    }

    fun isSourceTransactional(sourceId: String): Boolean {

        if (!sourceId.contains("/")) {
            return false
        }
        val appNameAndSourceId = sourceId.split('/', limit = 2)
        val sourceMetaId = appNameAndSourceId[0] + "/src@" + appNameAndSourceId[1]

        return sourceIdMeta.get(sourceMetaId).isTransactional
    }

    private fun completeTransaction(recordRefs: List<RecordRef>, action: TxnBody.TxnAction) {

        if (recordRefs.isEmpty()) {
            return
        }

        val context = RequestContext.getCurrentNotNull()

        var txnException: Exception? = null

        RecordsUtils.groupRefBySource(recordRefs).forEach { (sourceId, refs) ->

            val appName = sourceId.substringBefore("/", "")

            if (appName.isNotBlank() && isSourceTransactional(sourceId)) {
                val appRefs = refs.map { it.value }
                try {
                    commitImplInApp(appName, appRefs, action, context)
                } catch (e: Exception) {
                    log.error { "Exception while txn commit '${context.ctxData.txnId}'. Records: $appRefs" }
                    // main transaction already completed, and we should
                    // make as much remote commits or rollbacks as possible
                    if (txnException == null) {
                        txnException = e
                    } else {
                        txnException?.addSuppressed(e)
                    }
                }
            }
        }
        val finalException = txnException
        if (finalException != null) {
            throw finalException
        }
    }

    private fun commitImplInApp(
        appName: String,
        refs: List<RecordRef>,
        action: TxnBody.TxnAction,
        context: RequestContext
    ) {

        val body = TxnBody()
        body.setRecords(refs.map { it.removeAppName() })
        body.setAction(action)
        setContextProps(body, context)

        var exception: Exception? = null
        for (i in 1..4) {
            try {
                exchangeRemoteRequest(appName, TXN_PATH, body, TxnResp::class, context)
                if (exception != null) {
                    log.info {
                        "$action request with txn ${body.txnId} app $appName and records ${body.records} " +
                            "was completed successfully after ${i - 1} retry"
                    }
                    exception = null
                }
                break
            } catch (e: Exception) {
                exception = e
                if (i == 4) {
                    break
                }
                val sleepTime = i * 1000L
                log.warn {
                    "$action request with txn ${body.txnId} app $appName and records ${body.records} " +
                        "failed with exception ${e::class.simpleName} " +
                        "msg: ${e.message}. Retry sleep: ${sleepTime}ms"
                }
                Thread.sleep(sleepTime)
            }
        }
        if (exception != null) {
            throw exception
        }
    }

    private fun evalSourceIdMeta(sourceMetaId: String): RecSrcMeta {
        val atts = recordsService.getAtts(RecordRef.valueOf(sourceMetaId), RecSrcMetaAtts::class.java)
        return RecSrcMeta(atts.isTransactional ?: false)
    }

    private fun setContextProps(body: RequestBody, ctx: RequestContext) {
        val ctxData = ctx.ctxData
        body.msgLevel = ctxData.msgLevel
        body.requestId = ctxData.requestId
        body.txnId = ctxData.txnId
        body.setRequestTrace(ctxData.requestTrace)
    }

    private fun <T : RequestResp> exchangeRemoteRequest(
        appName: String,
        requestPath: String,
        body: RequestBody,
        respType: KClass<T>,
        context: RequestContext
    ): T {

        val result = webClient.execute(appName, requestPath, body, respType.java).get()

        throwErrorIfRequired(result.messages, context)
        context.addAllMsgs(result.messages)
        txnActionManager.execute(result.txnActions, context)

        return result
    }

    private fun formatObjForLog(obj: Any): String {
        return Json.mapper.toString(
            if (obj is HasSensitiveData<*>) {
                obj.withoutSensitiveData()
            } else {
                obj
            }
        ) ?: "null"
    }

    private fun throwErrorIfRequired(messages: List<ReqMsg>, context: RequestContext) {

        if (context.ctxData.omitErrors) {
            return
        }

        for (idx in messages.size - 1 downTo 0) {
            val msg = messages[idx]
            if (msg.level == MsgLevel.ERROR) {
                val textMessage = when (msg.type) {
                    RecordsError.MSG_TYPE -> msg.msg.getAs(RecordsError::class.java)?.msg ?: msg.msg.asText()
                    else -> msg.msg.asText()
                }
                throw RemoteRecordsException(msg, textMessage)
            }
        }
    }

    fun getSourceInfo(sourceId: String): RecordsSourceMeta? {
        // todo
        return null
    }

    fun getSourcesInfo(): List<RecordsSourceMeta> = emptyList()

    fun setRecordsService(recordsService: RecordsService) {
        this.recordsService = recordsService
    }

    private data class RecSrcMeta(
        val isTransactional: Boolean
    )

    data class RecSrcMetaAtts(
        @AttName("features.transactional")
        val isTransactional: Boolean?
    )
}
